### PR TITLE
CI changes

### PR DESCRIPTION
**Pull request by timotei-litespeed**
Added PHP Code sniffer integration

For local:
- in Visual code install plugin: [PHP Sniffer](https://marketplace.visualstudio.com/items?itemName=wongjn.php-sniffer) 
- Plugin settings:
     - `Auto detect` set to Checked
     - Optional: can be set to run `on save` 

Prerequirement: Node 16, Composer, PHP 7.2
Install:
  make sure there is a empty folder: `.cache` in root of plugin folder
  open terminal
  run: `npm install`
  run: `composer install`
  to run sniffer manually: `composer run sniff-check`

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+	"require-dev": {
```
Comment (szepeviktor) [inline composer.json:2]: Please add "name": "litespeedtech/lscache_wp",

Comment (timotei-litespeed) [reply composer.json:3]: @szepeviktor thank you! added

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -8,19 +8,48 @@ on:
...
           node-version: '20'
           check-latest: true
           cache: npm
           cache-dependency-path: package-lock.json
 
-      - run: npm ci
+      # Make sure .cache folder exists
+      - name: 'Create folder .cache s'
+        run: |
+          mkdir -p .cache && cd .cache && touch .gitkeep
```
Comment (tynanbe) [inline .github/workflows/ci.yml:50]: Why is this needed? Isn't the `.cache` dir already created by Composer above? PHPCS might also create it.

File: phpcs.xml.dist [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	<!-- Remove unwanted prefix from filepaths -->
+	<arg name="basepath" value="./"/>
+
+	<!-- Check max 20 files in parallel -->
+	<arg name="parallel" value="20"/>
+
+	<!-- Show sniff codes in all reports -->
+	<arg value="ps"/>
+	
+	<arg name="parallel" value="8"/>
```
Comment (tynanbe) [inline phpcs.xml.dist:32]: Remove this and update L26-27 instead.

File: phpcs.xml.dist [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+	<config name="testVersion" value="7.2-"/>
+
+	<rule ref="PHPCompatibility">
+		<!-- Exclude false positives -->
+		
+		<!-- PHP_INT_MIN is defined by WordPress if needed -->
+		<!-- <exclude name="PHPCompatibility.Constants.NewConstants.php_int_minFound" /> -->
+		<!-- hash_equals is defined by WordPress if needed -->
+		<!-- <exclude name="PHPCompatibility.FunctionUse.NewFunctions.hash_equalsFound" /> -->
```
Comment (tynanbe) [inline phpcs.xml.dist:65]: These were set up when testing with PHP v5.3. If they're not needed anymore, just remove them.

File: phpcs.xml.dist [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		<!-- array_key_firstFound is defined in plugin/lib -->
```
Comment (tynanbe) [inline phpcs.xml.dist:66]:
```xml
<!-- array_key_firstFound is defined in lib/php-compatibility.func.php -->
```
Please.

File: phpcs.xml.dist [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	#############################################################################
+	FILE SELECTION
+	Set which files will be subject to the scans executed using this ruleset.
+	#############################################################################
+	-->
+
+	<file>.</file>
+
+	<!-- Exclude any wordpress folder in the current directory -->
+	<!-- <exclude-pattern type="relative">^wordpress/*</exclude-pattern> -->
```
Comment (tynanbe) [inline phpcs.xml.dist:44]: I had this off for false positive testing. Change this line to just this, please
```xml
<exclude-pattern type="relative">^wordpress/*</exclude-pattern>
```

This line allows keeping a WP install in a local checkout of `lscache_wp` for better language server support across different editors.

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+	"prefer-stable": true,
+	"scripts": {
+		"post-install-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
+		],
+		"post-update-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
+		],
+		"sniff-check": [
+			"./vendor/bin/phpcs -p . --standard=./phpcs.xml.dist"
```
Comment (tynanbe) [inline composer.json:16]: This line should be fine as simply, correct?
```json
			"vendor/bin/phpcs"
```

File: composer.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+		"squizlabs/php_codesniffer": "*",
+		"phpcompatibility/php-compatibility": "*"
+	},
+	"prefer-stable": true,
+	"scripts": {
+		"post-install-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
+		],
+		"post-update-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
```
Comment (tynanbe) [inline composer.json:13]:
```json
			"vendor/bin/phpcs --config-set installed_paths vendor/phpcompatibility/php-compatibility"
```

File: composer.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
+	"name": "litespeedtech/lscache_wp",
+	"require-dev": {
+		"squizlabs/php_codesniffer": "*",
+		"phpcompatibility/php-compatibility": "*"
+	},
+	"prefer-stable": true,
+	"scripts": {
+		"post-install-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
```
Comment (tynanbe) [inline composer.json:10]:
```json
			"vendor/bin/phpcs --config-set installed_paths vendor/phpcompatibility/php-compatibility"
```

File: .gitignore [hunk 10]
```diff
@@ -6,3 +6,4 @@
 .env*
 .vscode
 node_modules
+/vendor
```
Comment (tynanbe) [inline .gitignore:9]: There should be a newline at end of file.

File: composer.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+		],
+		"post-update-cmd": [
+			"\"vendor/bin/phpcs\" --config-set installed_paths vendor/phpcompatibility/php-compatibility"
+		],
+		"sniff-check": [
+			"./vendor/bin/phpcs -p . --standard=./phpcs.xml.dist"
+		]
+	}
+}
```
Comment (tynanbe) [inline composer.json:19]: There should be a newline at end of file.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -8,19 +8,48 @@ on:
...
   pull_request:
 
+# Cancels all previous workflow runs for pull requests that have not completed.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.sha }}
+  cancel-in-progress: true
+
 jobs:
-  test:
+  checks:
```
Comment (tynanbe) [inline .github/workflows/ci.yml:17]:
```yml
  check:
```

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -8,19 +8,48 @@ on:
...
+      # Initialize Cache Composer
+      - name: Cache Composer dependencies
+        uses: actions/cache@v3
+        with:
+          path: .cache
+          key: ${{ runner.os }}-${{ hashFiles('**/composer.lock') }}
+
+      # Initialize Composer
+      # Will throw error to upgrade to Composer V2.
+      # We stick to V1 because of PHP 7.3 usage. https://github.com/marketplace/actions/composer-update-action
```
Comment (tynanbe) [inline .github/workflows/ci.yml:31]:
```yml
      # Initialize Composer
      # Use v1; v2 throws an error due to PHP v7.2 support
      # Ref: https://github.com/marketplace/actions/composer-update-action
```

File: phpcs.xml.dist [hunk 14]
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0"?>
+<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="LiteSpeed_Cache" xsi:noNamespaceSchemaLocation="lsc.xsd">
```
Comment (tynanbe) [inline phpcs.xml.dist:2]: Why change from this?
```xml
<ruleset name="LiteSpeed Cache Coding Standards">
```

File: phpcs.xml.dist [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+	<exclude-pattern>/node_modules/*</exclude-pattern>
+	<exclude-pattern>/vendor/*</exclude-pattern>
+
+
+	<!--
+	#############################################################################
+	SET UP THE RULESET
+	#############################################################################
+	-->
+
```
Comment (tynanbe) [inline phpcs.xml.dist:56]: Let's add a comment here:
```xml
	<!-- Check PHP v7.2 and all newer versions -->
```

**Review by tynanbe — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -8,19 +8,48 @@ on:
       - 'v*.*'
   pull_request:
 
+# Cancels all previous workflow runs for pull requests that have not completed.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.sha }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (timotei-litespeed) [reply .github/workflows/ci.yml:14]: Commented the code
Scenario was if you have more than 1 push in almost the same time :)

File: .github/workflows/ci.yml (hunk 2 above)
Comment (timotei-litespeed) [reply .github/workflows/ci.yml:50]: Is needed in case the folder do not exist on repository

File: .github/workflows/ci.yml (hunk 12 above)
Comment (timotei-litespeed) [reply .github/workflows/ci.yml:17]: Done

File: .github/workflows/ci.yml (hunk 13 above)
Comment (timotei-litespeed) [reply .github/workflows/ci.yml:31]: Changed

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -8,19 +8,48 @@ on:
...
+        env:
+          COMPOSER: 'composer.json'
+        with:
+          php_version: '7.2'
+          version: 1
+
+      # Initialize Node.js
       - uses: actions/setup-node@v3
         with:
           node-version: '20'
```
[the remark this answers is not in the document]
Comment (timotei-litespeed) [reply .github/workflows/ci.yml:42]: Changed version for node

File: .gitignore (hunk 10 above)
Comment (timotei-litespeed) [reply .gitignore:9]: Added

File: composer.json (hunk 9 above)
Comment (timotei-litespeed) [reply composer.json:10]: changed

File: composer.json (hunk 8 above)
Comment (timotei-litespeed) [reply composer.json:13]: changed

File: composer.json (hunk 11 above)
Comment (timotei-litespeed) [reply composer.json:19]: It get removed by Prettier. Should I add it manually?

File: phpcs.xml.dist (hunk 14 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:2]: Was added when I tried to fix a bug. But was from somwehere else.
Copied from other repositories.
I reverted the changes.

File: phpcs.xml.dist (hunk 3 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:32]: My mistake. I did not saw it's duplicated. Left only one line

File: phpcs.xml.dist (hunk 6 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:44]: Changed

File: phpcs.xml.dist (hunk 15 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:56]: Added

File: phpcs.xml.dist (hunk 4 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:65]: Removed

File: phpcs.xml.dist (hunk 5 above)
Comment (timotei-litespeed) [reply phpcs.xml.dist:66]: Changed

**Comment by szepeviktor** [comment]
I think this is not the GHA setup you want.
May I contribute an improved strategy after this one gets merged?

**Comment by timotei-litespeed** [comment]
Contribute now and we can see now, before merge

**Comment by szepeviktor** [comment]
> Contribute now, before merge

I have to inform you that I am  struggling with

- lack of EditorConfig (for Prettier)
- lack of syntax error check
- lack of coding standard
- lack of static analysis
- all right, we don't have unit/E2E tests

**Comment by timotei-litespeed** [comment]
@szepeviktor 
Ok, No problem.
Do you have any suggested changes to Github CI actions?

For us current script is almost enought, we added Code Sniffer as a code check, no to correct or fix the code.

**Comment by szepeviktor** [comment]
> Do you have any suggested changes to Github CI actions?

Yes. All the above tools. e.g. #696

File: composer.json (hunk 11 above)
Comment (tynanbe) [reply composer.json:13]: No, that's fine.

**Comment by timotei-litespeed** [comment]
Moved to https://github.com/litespeedtech/lscache_wp/pull/744

@szepeviktor thank you for the code and iddeeas
We are not really using WP code style and we are not adding Code Beautify(at least for now there are no plans)

**Comment by szepeviktor** [comment]
@timotei-litespeed Without the above tools, or without a coding standard you are theoretically cut off from bug-free code. That is very sad to see.

**Comment by timotei-litespeed** [comment]
I agree that withput a coding standard there is posibility for bugs to appear. For now this automation will help us make the code cleaner.

Thank you for suggestions but for NOW the changes are enought for the time allocated for this change.

👍
